Selecting number of batches from training and validation to use from original training set

kothasuhas opened the pull request:
Adding the ability to specify
- a number of batches to cut off the train dataset to add as an iid val dataset (`max_batches_dict`)
- a maximum number of batches to take from the train dataset (useful when u want to do epoching experiments as a function of token count rather than dataset size) (`num_validation_batches_dict`)

Resolves https://github.com/stanford-crfm/marin/issues/708

dlwh reviewed (commented):
I dunno, i don't love this. 

I get it's important to be able to do experiments like this but I'm getting a bit wary of the amount of special code paths in text.py

I also don't love the state stuff.

kothasuhas commented:
I agree this isn't the prettiest fix. It is definitely very helpful for my experiments to do this: rebuilding every dataset to have a val split is a lot of work and not as simple as partitioning the shards (example here: https://github.com/stanford-crfm/marin/issues/708)

Did you have any suggestions for how to best handle this? Here are possible routes
- Try to do this on an individual dataset level instead of mixture: the barrier here is that this would have to be at the cache level, which iiuc doesnt permit the niceness of ds.slice_dataset
- Do this from marin side: should I try to write a levanter util that slices the components external to this, and then have marin call that utility?

kothasuhas commented on src/levanter/data/text.py:
```diff
@@ -1125,6 +1125,12 @@ class LMMixtureDatasetConfig(LMTaskConfig):
     of samples, equal to the expected number of samples in the mixture, rounding in the expected way."""
 
+    max_batches_dict: Optional[Dict[str, int]] = None
+    """ Maximum number of batches to use from each dataset for training (using the initial batch size)"""
```
it does not support the precise operation of taking a fixed number of tokens as a subset and epoching on this. for example, you can't easily simulate "what does 4 epochs of 3B DCLM tokens" looks like, and doing 4 epochs of all of DCLM is too expensive

kothasuhas commented on src/levanter/data/text.py:
```diff
@@ -1125,6 +1125,12 @@ class LMMixtureDatasetConfig(LMTaskConfig):
+    max_batches_dict: Optional[Dict[str, int]] = None
+    """ Maximum number of batches to use from each dataset for training (using the initial batch size)"""
+
+    num_validation_batches_dict: Optional[Dict[str, int]] = None
```
breaking out a subset is annoying :// and not possible for datasets like starcoder

dlwh replied:
can we actually just rename this num_validation_batches? we don't usually put type in the name of the param

kothasuhas commented on src/levanter/data/text.py:
```diff
@@ -1234,13 +1251,45 @@ def shuffle_ds(ds, key):
                 sliced_datasets[name] = ds.slice_dataset(end_index=simulated_length_of_dataset)
             datasets = sliced_datasets
 
+        self.validation_token_datasets = {}
```
fixed to not mutate state, thanks!

kothasuhas commented:
addressed comments, only suspicious choice is having a flag "validation_batch_size" set to a default of 1024. i think its a little ugly to change the validation signature and pass in the batch schedule, only to get a validation batch size which is almost never important

dlwh commented on src/levanter/data/text.py:
```diff
@@ -1234,13 +1252,38 @@ def shuffle_ds(ds, key):
+            for name, ds in datasets.items():
+                if name in self.num_validation_batches_dict:
+                    num_sequences = self.num_validation_batches_dict[name] * self.validation_batch_size
+                    len_dataset = len(ds.as_sync_dataset())
```
it'd be more async-friendly if we just did validation=slice_dataset(end_index=num_sequences) and training=slice_dataset(start_index=num_sequences)

kothasuhas replied:
done!

kothasuhas commented:
dont merge yet, need to fix something first

kothasuhas commented:
two updates 
1. there was a bug in the prior version when removing the mutation since i was not reusing the train caches for the train datasets. this has been fixed now
2. for some reason, the suggested change to make the dataset more async friendly (splitting at num_sequences instead of len - num_sequences) does not interact well with epoching results in `raise ValueError("Requested indices beyond the end of the dataset")`. i have a hunch it's due to how `EpochDataset` interacts with `start_index`. for now, i don't think the overhead is very high to query length and it works so i've reverted to the prior version

i'm going to do some more testing with some actual training runs to make sure things are well-behaved before merging

kothasuhas commented:
quick note: when i test this script, it always works with no lm eval harness. when i tested with lm eval harness, it often crashed, but i think thats related to us central 2 + harness since this PR shouldn't touch that. just sharing for visibility

dlwh commented on src/levanter/data/text.py:
```diff
@@ -1153,12 +1179,17 @@ def train_set(
         if isinstance(weights, Sequence):
             weights = rescale_mixture_schedule_for_batch_schedule(weights, batch_schedule)
 
+        initial_batch_size = batch_schedule.batch_size_at_step(0)
```
really probably the right thing is to use EvalBatch instead but this is fine

dlwh reviewed (approved):
i think this is fine but if you want to throw it behind an lru_cache or something that would fix
